### PR TITLE
Extract similar concerns from the torch data loader into a `TorchDataSource`.

### PR DESCRIPTION
**Pull request by alxmrs**
A drive by refactor made during #532. I noticed that the means/stds/mask arrays really should be grouped with their associated tensor (either input, label, or boundary). The interface mirrors `DataSource`, except all operations are on `torch.Tensor`s rather than Xarray `Dataset`s. 

As a sanity check, I ran a local performance profile before and after this change. On my machine, both take basically the same median amount of time.

### CODE REVIEW

File: src/ocean_emulators/utils/data.py [hunk 1]
```diff
@@ -299,6 +299,59 @@ def from_datasets(
...
 
 
+@dataclasses.dataclass
+class TorchDataSource:
+    data: Float[torch.Tensor, "batch time variable lat lon"]
+    means: Float[torch.Tensor, " variable"]
+    stds: Float[torch.Tensor, " variable"]
+    mask: Bool[torch.Tensor, " variable"]
+
+    def map(self, data_func: Callable) -> Self:
```
Comment (alxmrs) [inline src/ocean_emulators/utils/data.py]: Instead of `map`, this could be `slice` or `__getitiem__` where we would allow the user to pass the slicing information to get the data before or after `hist`.

File: src/ocean_emulators/datasets.py [hunk 2]
```diff
@@ -470,11 +474,12 @@ def __init__(
             indices_da + stride * window_dim
         )
 
-        self.wet_input: PrognosticMask = src.masks.prognostic.to(self.device)
-        self.wet_surface: GridMask = src.masks.boundary.to(self.device)
-        self.wet_label: PrognosticMask = dst.masks.prognostic.to(self.device)
+        # NB(alxmrs): Keep masks on CPU - will be moved to GPU in to_train_data()
```
Comment (alxmrs) [inline src/ocean_emulators/datasets.py:483]: In my experiment branch, this was necessary to remove device related errors.

File: src/ocean_emulators/datasets.py [hunk 3]
```diff
@@ -483,16 +488,16 @@ def flatten_to_device(means_or_stds: xr.Dataset) -> torch.Tensor:
...
 
-        self.boundary_means = flatten_to_device(self._boundary_src.means)
-        self.boundary_stds = flatten_to_device(self._boundary_src.stds)
+        self.boundary_means = flatten(self._boundary_src.means)
+        self.boundary_stds = flatten(self._boundary_src.stds)
 
-        self.label_means = flatten_to_device(self._label_src.means)
-        self.label_stds = flatten_to_device(self._label_src.stds)
+        self.label_means = flatten(self._label_src.means)
+        self.label_stds = flatten(self._label_src.stds)
```
Comment (alxmrs) [inline src/ocean_emulators/datasets.py]: Question for the reviewer: should this be moved into the new class? e.g. like on post init?

File: src/ocean_emulators/utils/data.py [hunk 4]
```diff
@@ -299,6 +299,83 @@ def from_datasets(
...
+
+    @classmethod
+    def from_data_source(
+        cls,
+        data: Float[torch.Tensor, "batch time variable lat lon"],
+        mask: Float[torch.Tensor, " variable"],
+        src: DataSource,
+    ) -> Self:
+        means_torch = _flatten(src.means)
+        stds_torch = _flatten(src.stds)
```
Comment (alxmrs) [inline src/ocean_emulators/utils/data.py:349]: I did a number of local profiles, and I haven't found a consistent result that says that we need to cache these numbers in the calling class. If anything, I have found that this "cleaner" approach is slightly faster. Hard to say until we run automated benchmarks.

File: src/ocean_emulators/datasets.py (hunk 2 above)
Comment (jder) [reply src/ocean_emulators/datasets.py:483]: This is a little worrying from a performance perspective. We're now doing twice as many transfers to GPU. Probably doesn't matter too much but given it worked before I don't understand why it's a problem now.

**Comment by jder** [comment]
@codex review

File: src/ocean_emulators/utils/data.py [hunk 5]
```diff
@@ -299,6 +299,83 @@ def from_datasets(
...
+            "(variable lev)=var",
+            concat_dim="var",
+        ).rename({"var": "variable"})
+    else:
+        array = means_or_stds.to_dataarray()
+    return torch.from_numpy(array.to_numpy().flatten())
+
+
+@dataclasses.dataclass
+class TorchDataSource:
```
Comment (jder) [inline src/ocean_emulators/utils/data.py]: I generally like the idea of trying to give semantic information around our torch tensors, thanks for doing this! I don't think "DataSource" is the right name, personally. It is an actual slice of ocean data, just with some metadata. How about "TorchDataSlice" or "OceanTensor" or "DecoratedData" or something?

Either way, please add a docstring here explaining what it represents.

File: src/ocean_emulators/utils/data.py (hunk 1 above)
Comment (jder) [reply src/ocean_emulators/utils/data.py]: Yeah, I don't think this is a good abstraction right now -- it's easy to call map with a lambda whose result no longer matches the means/stds/mask. So either the return type here should be the raw data Tensor, or we should restrict this to only do things to data which we can guarantee maintains the invariant of this class (ie that all the variable information lines up across the 4 members)

**Review by jder — CHANGES_REQUESTED** [review]
Thanks! Generally in favor of this change. Just a few suggestions!

File: src/ocean_emulators/utils/data.py (hunk 5 above)
Comment (jder) [reply src/ocean_emulators/utils/data.py]: (Might also be nice to add a test or two)

File: src/ocean_emulators/utils/data.py (hunk 1 above)
Comment (alxmrs) [reply src/ocean_emulators/utils/data.py]: What do you think of the current `with_time` interface?

**Comment by jder** [comment]
@codex review

File: src/ocean_emulators/utils/data.py [hunk 6]
```diff
@@ -299,6 +299,104 @@ def from_datasets(
...
+        )
+
+    def pin_memory(self, device: torch.device | None = None) -> Self:
+        return dataclasses.replace(
+            self,
+            data=self.data.pin_memory(device=device),
+            means=self.means.pin_memory(device=device),
+            stds=self.stds.pin_memory(device=device),
+            mask=self.mask.pin_memory(device=device),
+        )
```
Comment (jder) [inline src/ocean_emulators/utils/data.py]: This seems to be unused? Also I don't think pin_memory takes a device?

File: src/ocean_emulators/utils/data.py [hunk 7]
```diff
@@ -299,6 +299,104 @@ def from_datasets(
...
+        """Normalize and mask tensors."""
+        tensor = self.data
+        if normalize_before_mask:
+            tensor = self._normalize(tensor)
+        tensor = torch.where(self.mask, tensor, masked_fill_value)
+        if not normalize_before_mask:
+            tensor = self._normalize(tensor)
+        return tensor
+
+    def to(self, device: torch.device, non_blocking: bool = False) -> Self:
```
Comment (jder) [inline src/ocean_emulators/utils/data.py]: 🐑 we always seem to call this with non_blocking=True; maybe make the argument mandatory or default to True?

File: src/ocean_emulators/utils/data.py (hunk 1 above)
Comment (jder) [reply src/ocean_emulators/utils/data.py]: Looks great to me, thanks!

**Review by jder — APPROVED** [review]
Thanks! Just two little things.

File: src/ocean_emulators/utils/data.py (hunk 6 above)
Comment (alxmrs) [reply src/ocean_emulators/utils/data.py]: I remember you once mentioning that some valuable part of pytorch checks if a class has a `to` and `pin_memory` method in order to do something that we needed. Do you remember that context? Is that even relevant here?

File: src/ocean_emulators/utils/data.py (hunk 7 above)
Comment (alxmrs) [inline src/ocean_emulators/utils/data.py]:
```suggestion
    def to(self, device: torch.device, non_blocking: bool = True) -> Self:
```

File: src/ocean_emulators/utils/data.py (hunk 6 above)
Comment (jder) [reply src/ocean_emulators/utils/data.py]: Yes, the things that come out of the DataLoader (TrainData) have `pin_memory()` called on them

Comment (alxmrs) [reply src/ocean_emulators/utils/data.py]: Ah, cool. So, since this never exits the loader, I'll delete this unused method. We can add it later if it's needed. Thanks.
